Cleaned up & updated CI readme (#13312)
GoogleCloudPlatform/magic-modules: .ci/README.md

magic-modules CI
===

## Playbook: Resolving downstream build failures

There are four downstreams, each of which has a corresponding "sync branch" in `magic-modules` which tracks the _most recently generated commit_ from magic-modules to that downstream.

- Downstream: [hashicorp/terraform-provider-google](https://github.com/hashicorp/terraform-provider-google)  
  Sync branch: [`tpg-sync`](https://github.com/GoogleCloudPlatform/magic-modules/tree/tpg-sync)
- Downstream: [hashicorp/terraform-provider-google-beta](https://github.com/hashicorp/terraform-provider-google-beta)  
  Sync branch: [`tpgb-sync`](https://github.com/GoogleCloudPlatform/magic-modules/tree/tpgb-sync)
- Downstream: [GoogleCloudPlatform/terraform-google-conversion](https://github.com/GoogleCloudPlatform/terraform-google-conversion)  
  Sync branch: [`tgc-sync`](https://github.com/GoogleCloudPlatform/magic-modules/tree/tgc-sync)
- Downstream: [terraform-google-modules/docs-examples](https://github.com/terraform-google-modules/docs-examples)  

>   Sync branch: [`tf-oics-sync`](https://github.com/GoogleCloudPlatform/magic-modules/tree/tpg-sync)  

Reviewer: ```suggestion
  Sync branch: [`tf-oics-sync`](https://github.com/GoogleCloudPlatform/magic-modules/tree/tf-oics-sync)  
```

  Note: `oics` refers to "Open in Cloud Shell".

The goal of this system is that each downstream commit will have exactly one MM commit that it corresponds to, and each MM commit will correspond to at most one commit in a downstream. If an MM commit had no changes in a downstream, no commit will be created.

The sync branches allow downstream generation for each downstream to wait until the previous commit for that downstream has finished generating. If downstream generation fails for one commit, the following commits will continue to wait for 24 hours.

Run the following command to verify what commits the sync branches are pointing to:

```
`git fetch origin && git rev-parse origin/tpg-sync origin/tpgb-sync origin/tf-oics-sync origin/tgc-sync`
```

### Transient GitHub failures

> Most downstream build failures are transient GitHub failures. To resolve these, click the "Retry" button in Cloud Build. This is safe because downstream builds are idempotent; if a commit has already been generated, we will not make a new commit.

Reviewer: > Most downstream build failures are transient GitHub failures.

Do 5XX errors encapsulate most of these?

Author: I think they can also be 4XX (for example auth errors); I'm not sure if one or the other is more common.

### Downstream build job is not triggered by commits.
This is rare but we've seen this happened before. In this case, we need to manually trigger a Cloud Build job by running 
```
gcloud builds triggers run build-downstreams --project=graphite-docker-images --substitutions=BRANCH_NAME=<BASE_BRANCH_NAME> --sha=<COMMIT_SHA>
```
You'll need to substitute `<COMMIT_SHA>` with the commit sha that you'd like to trigger the build against and `<BASE_BRANCH_NAME>` with base branch that this commit is pushed into, likely `main` but could be feature branches in some cases.

### Magician / generation bugs
Magician or generation bugs are extremely rare. They cause generation itself to fail loudly, and the commits that introduce them need to be skipped.

Be sure that you understand the mechanism for how the commit is causing a failure before proceeding. Skipping commits will cause multiple MM commits to be squashed into a single downstream commit, which breaks commit linking expectations and release notes.

> 1. Lock the magic-modules main branch

Reviewer: ```suggestion
1. Lock the magic-modules main branch (or ask the team to)
```

2. Create a PR to fix the bug
3. Unlock the main branch, submit the PR, and re-lock the main branch
4. Update the sync branches to point to the bad commit
   - This indicates that we've "already generated" the downstreams for this commit, so builds for the following commit will stop waiting
6. Verify that downstream generation succeeded - this should only take <15 minutes (even though the overall build-downstreams build takes ~1hr)
7. Unlock the main branch

### Manually pushing commits to downstreams
In general, this should not be necessary because this situation shouldn't come up. It is a historical process that may not work out of the box. The case where you might want to try this is:

- A commit broke the magician (or downstream generation)
- Multiple PRs were merged before the main branch was locked

In this case, skipping just the initial commit would not work (because the following commit doesn't contain a fix) and skipping all the commits is not desirable (because we don't want to squash them.) You would need to instead locally apply the CI fix and then "manually" replicate the work of downstream generation.

> When this happened the first time, Cameron and Nathan wrote this little shell snippet, which should do most of the work for you.  You will need to get the Magician's github token, either by generating a new one (be sure to clean up after yourself when done), by decrypting the value in .ci/gcb-push-downstream.yml as cloudbuild does, or by accessing the token in Google's internal secret store.

Reviewer: ```suggestion
Legacy fix (may no longer work):
When this happened the first time, the team wrote this little shell snippet, which might do most of the work for you.  You will need to get the Magician's github token, either by generating a new one (be sure to clean up after yourself when done), by decrypting the value in .ci/gcb-push-downstream.yml as cloudbuild does, or by accessing the token in Google's internal secret store.
```
